language/stdlib/modules/option.move

```diff
@@ -0,0 +1,119 @@
+    }
+
+    // Return true if `t` does not hold a value
+    public fun is_none<Element>(t: &T<Element>): bool {
+        Vector::is_empty(&t.vec)
+    }
+
+    // Return true if `t` holds a value
+    public fun is_some<Element>(t: &T<Element>): bool {
+        !Vector::is_empty(&t.vec)
+    }
+
+    // Return true if the value in `t` is equal to `e_ref`
+    // Always returns `false` if `t` does not hold a value
+    public fun contains<Element>(t: &T<Element>, e_ref: &Element): bool {
+        Vector::borrow(&t.vec, 0) == e_ref
+    }
+
+    // Return an immutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow<Element>(t: &T<Element>): &Element {
+        Vector::borrow(&t.vec, 0)
+    }
+
+    // TODO: not clear this API is a good idea
+    // Return a reference to the value inside `t` if it holds one
+    // Return `default_ref` if `t` does not hold a value
+    public fun borrow_default<Element>(t: &T<Element>, default_ref: &Element): &Element {
+        let vec_ref = &t.vec;
+        if (Vector::is_empty(vec_ref)) {
+            Vector::borrow(vec_ref, 0)
+        } else {
+            default_ref
+        }
+    }
+
+    // TODO: not clear we want both this and borrow_default
+    // Return the value inside `t` if it holds one
+    // Return `default` if `t` does not hold a value
+    public fun get_default<Element: copyable>(t: &T<Element>, default: Element): Element {
+        let vec_ref = &t.vec;
+        if (Vector::is_empty(vec_ref)) {
+            *Vector::borrow(vec_ref, 0)
+        } else {
+            default
+        }
+    }
+
+    // Convert the none option `t` to a some option by adding `e`.
+    // Aborts if `t` already holds a value
+    public fun fill<Element>(t: &mut T<Element>, e: Element) {
+        let vec_ref = &mut t.vec;
+        if (Vector::is_empty(vec_ref)) {
+            Vector::push_back(vec_ref, e)
+        } else {
+            abort(99)
+        }
+    }
+
+    // Convert a `some` option to a `none` by removing and returning the value stored inside `t`
+    // Aborts if `t` does not hold a value
+    public fun extract<Element>(t: &mut T<Element>): Element {
+        Vector::pop_back(&mut t.vec)
+    }
+
+    // Unpack `t` and return its contents
+    // Aborts if `t` does not hold a value
+    public fun destroy_some<Element>(t: T<Element>): Element {
+        let T { vec } = t;
+        let elem = Vector::pop_back(&mut vec);
+        Vector::destroy_empty(vec);
+        elem
+    }
+
+    // Unpack `t`
+    // Aborts if `t` holds a value
+    public fun destroy_none<Element>(t: T<Element>) {
+        let T { vec } = t;
+        Vector::destroy_empty(vec)
+    }
+
+    // ===== APIs that might be controversial + I am inclined to exclude ====
+
+    // Why controversial: do we want interior mutability for an option `o` given that you could
+    // just (1) reasign `o` if it's copyable or (2) extract, then fill?
```

Comment:
I think you want exposed mutability of the element 
Imagine something like 
```
o: &mut Option::T<Coin::T>;
...
Coin::withdraw(Option::borrow_mut(o), 10)
```
Without borrow mut that becomes
```
let coin = Option::extract(o);
Coin::withdraw(&mut coin, 10);
Option::fill(o, coin)
```

Reply:
Yeah, fair point. The case where the callee wants to mutate the thing inside of the option is a good motivation for this case. Will add this